vpr/src/route/bucket.cpp

```diff
@@ -0,0 +1,234 @@
+    heap_size_ = required_number_of_buckets * 2;
+
+    heap_ = (BucketItem**)vtr::realloc((void*)(heap_),
+                                       heap_size_ * sizeof(BucketItem*));
+    std::fill(heap_ + old_size, heap_ + heap_size_, nullptr);
+}
+
+void Bucket::verify() {
+    for (size_t bucket = heap_head_; bucket <= heap_tail_; ++bucket) {
+        for (BucketItem* data = heap_[bucket]; data != nullptr;
+             data = data->next_bucket) {
+            VTR_ASSERT(data->item.cost > 0 && ((size_t)cost_to_int(data->item.cost)) == bucket);
+        }
+    }
+}
+
+void Bucket::empty_heap() {
+    if (heap_head_ != std::numeric_limits<size_t>::max()) {
+        std::fill(heap_ + heap_head_, heap_ + heap_tail_ + 1, nullptr);
+    }
+    heap_head_ = std::numeric_limits<size_t>::max();
+    heap_tail_ = 0;
+
+    // Quickly reset all items to being free'd
+    items_.clear();
+}
+
+void Bucket::check_scaling() {
+    float min_cost = min_cost_;
+    float max_cost = max_cost_;
+    VTR_ASSERT(max_cost != std::numeric_limits<float>::min());
+    if (min_cost == std::numeric_limits<float>::max()) {
+        min_cost = max_cost;
+    }
+    auto min_bucket = cost_to_int(min_cost);
+    auto max_bucket = cost_to_int(max_cost);
+
+    if (min_bucket < 0 || max_bucket < 0 || max_bucket > 1000000) {
+        // If min and max are close to each other, assume 3 orders of
+        // magnitude between min and max.
+        //
+        // If min and max are at least 3 orders of magnitude apart, scale
+        // soley based on max cost.
+        conv_factor_ = 50000.f / max_cost_ / std::max(1.f, 1000.f / (max_cost_ / min_cost_));
+
+        VTR_ASSERT(cost_to_int(min_cost_) >= 0);
+        VTR_ASSERT(cost_to_int(max_cost_) >= 0);
+        VTR_ASSERT(cost_to_int(max_cost_) < 1000000);
+
+        // Reheap after adjusting scaling.
+        if (heap_head_ != std::numeric_limits<size_t>::max()) {
+            std::vector<BucketItem*> reheap;
+            for (size_t bucket = heap_head_; bucket <= heap_tail_; ++bucket) {
+                for (BucketItem* item = heap_[bucket]; item != nullptr; item = item->next_bucket) {
+                    reheap.push_back(item);
+                }
+            }
+
+            std::fill(heap_ + heap_head_, heap_ + heap_tail_ + 1, nullptr);
+            heap_head_ = std::numeric_limits<size_t>::max();
+            heap_tail_ = 0;
+
+            for (BucketItem* item : reheap) {
+                push_back(&item->item);
+            }
+        }
+    }
+}
+
+void Bucket::push_back(t_heap* hptr) {
+    float cost = hptr->cost;
+    if (!std::isfinite(cost)) {
+        return;
+    }
+
+    bool check_scale = false;
+    // Exclude 0 cost from min_cost to provide useful scaling factor.
+    if (cost < min_cost_ && cost > 0) {
+        min_cost_ = cost;
+        check_scale = true;
+    }
+    if (cost > max_cost_) {
+        max_cost_ = cost;
+        check_scale = true;
+    }
+
+    if (check_scale) {
```

Comment:
Explain that you check_scaling (number of buckets) if the min_cost,max_cost range has changed, to ensure you've got a reasonable number of buckets.

Reply:
Done